### ISSUE TITLE
Add spacing to card numbers

### ISSUE THREAD

**Issue by bendrucker** [bendrucker/angular-credit-cards#18]
@bbean86 Care to weigh in on how thoroughly you've tested your implementation? I specifically left out any formatting because it adds a ton of complexity. You have to worry about data being pasted in and where the cursor's positioned as users are typing. In order to support this I'd have to get a suite of tests written up with Protractor to even consider releasing it. Before I spend that time I'd like to hear whether you've encountered any issues to date.

**Comment by bbean86** [comment]
There is a sore lack of unit testing on this feature on my part, and I apologize for that. However, I have monkey-tested this to death, as well as this going through a full cycle of QA at my company with no issues for Amex, MasterCard, Visa, and Discover card numbers. Pasting works, cursor positioning while typing is fine (no extra spaces or weirdness like that). It does rely on the `length` property for a card, though. Maybe that property should be renamed `maxLength`? Also, I honestly only did cursory research on the non-major card number formats, so I just gave them the xxxx xxxx xxxx xxxx format if I couldn't find anything definitive. We may want to do a little more digging there.

**Comment by bbean86** [comment]
Another thing to bring up is that for this to work in my implementation, on the input you have to use `ng-model-options='{ allowInvalid: true }'`. We could hide that behind a directive to turn this feature on, like `cc-formatted`. That directive would require `cc-number` and just apply the needed `ng-model-options`.

**Comment by bendrucker** [comment]
Well the bigger issue is that unit testing won't cut it on something like this. I have to spin up Sauce and write Protractor tests to verify this behavior.

**Comment by bendrucker** [comment]
Using `ngModelOptions` should be fine as a hack but it's not the right solution. This one will take some time because I want to be 100% sure that it's not introducing bugs with all the many edge cases that modifying user inputs on the fly can create. There's definitely value in it because it's way easier to parse what you're seeing on screen when it's spaced and you can chunk the numbers.

**Comment by atav32** [comment]
The formatting's pretty important in a professional product. It's the main reason that I'm going to fork laurihy/angular-payments and fix the bugs instead.

**Comment by bendrucker** [comment]
I don't have time to do this wrong, answer issues about it, and then hack on fixes over time. As advertised, I'm doing this right or not at all. I'd certainly welcome a PR here that handles it which I genuinely believe would take less time than trying to fix angular-payments.

**Comment by bendrucker** [comment]
This is actually super easy if you don't care about cursor bugs (3dd5bd12d766758bbdd8ef5681906744537bec33)

I want to at least take a crack at addressing the fact that any keystroke results in the cursor being moved to the end of the input. I can't dive too deep this weekend so if I can't resolve it I'll still ship a new minor version with formatting disabled by default and enabled via an attribute. I feel a responsibility to do this right but if Stripe's willing to ship buggy forms for these rare cases I'm open to allowing opt in. 

![Stripe ships bugs](http://f.cl.ly/items/3K00132w0u0o2k2m3T1W/Screen%20Recording%202015-03-29%20at%2004.30%20PM.gif)

**Comment by bendrucker** [comment]
I pushed this in 2.3.0. There are some easily reproducible cursor positioning bugs, but that's also true of every single other open source implementation I could find. I picked the cases that seemed most unlikely to ignore for now and focused on getting obvious/common stuff right. It's also got the advantage of not relying on DOM events at all, so none of the nonsense from angular-payments with short-circuiting Angular's input handling and causing all kinds of obscure bugs. 

I'm basically going to leave this as is for the foreseeable unless anyone is up for doing PRs. If and when I get time to work on input masking again it'll be in the form of a super thorough vanilla implementation that can be plugged in here. I feel like anything in between isn't worth the effort.
